lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -4162,6 +4162,30 @@ Optional<ProtocolConformanceRef> TypeChecker::conformsToProtocol(
   return lookupResult;
 }
 
+// External API for checking protocol conformance outside the TypeChecker.
+//
+// Note: Much of the TypeChecker::conformsToProtocol and
+// TypeChecker::checkGenericArguments functionality does not
+// make sense in this setting. We need to reason about dead code paths as
+// follows:
+//
+// - ConformanceCheckFlags skips dependency checking.
+//
+// - Passing an invalid SourceLoc skips diagnostics.
+//
+// - Type::subst will be a nop, because 'source' type is already fully
+// substituted in SIL. Consequently, TypeChecker::LookUpConformance, which
+// is only valid during type checking, will never be invoked.
+//
+// - mapTypeIntoContext will be a nop.
+Optional<ProtocolConformanceRef>
+ModuleDecl::conformsToProtocol(Type sourceTy, ProtocolDecl *targetProtocol) {
```

Comment:
I'm not really happy with how this is implemented. Was it too difficult to untangle the conditional requirement checking from conformsToProtocol/checkGenericArgs() into something that can live in the ModuleDecl?

Reply:
I'm thrilled that I could get the functionality that we need for 5.0 without rewriting part of the TypeChecker.

Which parts are you unhappy about:

The code in this PR itself? I don't see a downside to simply exposing the TypeChecker API like this in the short term.

The TODO about the TypeChecker only reporting the positive (WillSucceed) case and not also reporting the negative case? Changing that will introduce a lot of unnecessary risk in the short term, and frankly it's beyond my skill level.

The lengthy "Note" comment above about dead code paths? I toned down the language, but some explanation is warranted. Otherwise, it's hard to tell just by reading the `TypeChecker::conformsToProtocol` implementation how it works in different contexts. It's controlled by flags and callbacks.

Are you or unhappy with the existing `conformsToProtocol` implementation? We'll see what @DougGregor says, but I think he wants the TypeChecker interface to be directly callable from the cast optimizer as a fully encapsulated API. That means there is a fair amount of dead code behind the TypeChecker interface, depending on how it's called. This is usually a sign of poor design, which is why I added the lengthy note. On the other hand, it's a natural tradeoff if we're striving for a fully general TypeChecker API.

When I first looked at this, I wanted to factor out the requirement checking part, so the code would be easier to understand. Then I realized two things
- Most of the complexity is really needed to handle recursive requirement checking, even though a lot of the book-keeping is irrelevant.
- @DougGregor 's vision here is for the TypeChecker API to be callable at any point as if it's a separate subsystem. Correct me if I'm wrong.

Reply:
I think the conformsToProtocol() API is not very good, yeah. I guess this change is fine for 5.0 but we should re-think this whole interface.

Reply:
Yeah, there are two refactorings to be done here, neither of which is appropriate for the 5.0 branch:

(1) Turn the "conforms to protocol" check into a request and separate it from the type checker instance, and
(2) Generalize the logic of the "check generic arguments" to report pass/fall/may succeed separately